### PR TITLE
services/horizon/expingest: State machine tests

### PR DESCRIPTION
**Pull request by bartekn**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

This commit is adding tests to each state of ingestion state machine.

### Why

We need to add tests to increase test coverage of Horizon's ingestion system. While working on the tests several issues has been found:
* Off-by-one bug in `resume` state (see `TestIngestOrderbookOnlyWhenLastLedgerExpEqualsCurrent` test for more info).
* Transition to invalid state in `init` when last ingested ledger > 0 but latest history ledger = 0 (see `TestResumeStateBehindHistory0` for more details).
* `graph.Apply` was called with invalid value in `loadOffersIntoMemory`.
* Added the code to update cursor in stellar-core (and take the cursor name from config).
* Removed waiting between `resume` transitions when catching up (see `noSleep` state variable).
* Simplified returning error in `run()` method in `shutdown` state.
* Extracted `ProcessorsRunner` to simplify testing.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -10,7 +10,19 @@
             "port": 2345,
             "host": "127.0.0.1",
             "program": "${workspaceRoot}/services/horizon/main.go",
-            "env": {},
+            "env": {
```
Comment (tamirms) [inline .vscode/launch.json]: I think this was mistakenly included in the PR

File: services/horizon/internal/expingest/processors_runner.go [hunk 2]
```diff
@@ -18,63 +22,36 @@ const (
...
-type groupTransactionProcessors []horizonTransactionProcessor
-
-func (g groupTransactionProcessors) ProcessTransaction(tx io.LedgerTransaction) error {
-	for _, p := range g {
-		if err := p.ProcessTransaction(tx); err != nil {
-			return err
-		}
-	}
-	return nil
+type ProcessorsRunnerInterface interface {
```
Comment (tamirms) [inline services/horizon/internal/expingest/processors_runner.go]: I think ProcessorRunnerInterface is a better name

File: services/horizon/internal/expingest/processors_runner.go [hunk 3]
```diff
@@ -18,63 +22,36 @@ const (
...
-func (g groupTransactionProcessors) Commit() error {
-	for _, p := range g {
-		if err := p.Commit(); err != nil {
-			return err
-		}
-	}
-	return nil
+var _ ProcessorsRunnerInterface = (*ProcessorsRunner)(nil)
+
+type ProcessorsRunner struct {
```
Comment (tamirms) [inline services/horizon/internal/expingest/processors_runner.go]: I prefer ProcessorRunner

File: services/horizon/internal/expingest/processors_runner.go [hunk 4]
```diff
@@ -178,7 +153,7 @@ func (s *System) runHistoryArchiveIngestion(
 			s.historyArchive,
 			&io.MemoryTempSet{},
 			checkpointLedger,
-			s.maxStreamRetries,
+			s.config.MaxStreamRetries,
```
Comment (tamirms) [inline services/horizon/internal/expingest/processors_runner.go]: maybe include `maxStreamRetries` in ProcessorsRunner so we can remove the dependency on config?

Comment (tamirms) [reply services/horizon/internal/expingest/processors_runner.go]: nevermind, I see config is used in other places

File: .vscode/launch.json (hunk 1 above)
Comment (abuiles) [reply .vscode/launch.json]: fixed

File: services/horizon/internal/expingest/main.go [hunk 5]
```diff
@@ -899,6 +988,26 @@ func (s *System) resetStateVerificationErrors() {
 	s.stateVerificationErrors = 0
 }
 
+func (s *System) updateCursor(ledgerSequence uint32) error {
```
Comment (tamirms) [inline services/horizon/internal/expingest/main.go]: what does this do?

Comment (bartekn) [reply services/horizon/internal/expingest/main.go]: It updates the cursor in stellar-core with the latest ingested ledger sequence so stellar-core knows it's safe to remove the ledgers from it's DB. Without it, it's possible that core removes the ledger before it's processed by Horizon. See: https://www.stellar.org/developers/stellar-core/software/commands.html#http-commands

**Comment by tamirms** [comment]
overall, looks good! I have two comments:

are there other places where we want to skip sleep? The transitions I have in mind are going from init to build, init to resume, init to ingestHistoryRange, build to resume, ingestHistoryRange to init (when there are no errors). It seems like we want to skip sleep if we transition states with no errors.

because it is possible to go back to init from resume, I think we should be clearing the orderbook graph (e.g. removing all offers) at the beginning of `loadOffersIntoMemory()` and `buildState()`. Otherwise there will be errors when applying orderbook graph batches.

**Comment by bartekn** [comment]
> because it is possible to go back to init from resume, I think we should be clearing the orderbook graph (e.g. removing all offers) at the beginning of loadOffersIntoMemory() and buildState(). Otherwise there will be errors when applying orderbook graph batches.

Good catch! I added `Clear()` method that removes all offers from the graph and call it in `loadOffersIntoMemory`.

> are there other places where we want to skip sleep? The transitions I have in mind are going from init to build, init to resume, init to ingestHistoryRange, build to resume, ingestHistoryRange to init (when there are no errors). It seems like we want to skip sleep if we transition states with no errors.

I think we can do it but let's do it in a separate PR. My only worry is that in the future we'll introduce transition between two states without errors and we'll be flooded with the logs.

**Pull request merged**